### PR TITLE
(CDAP-14646) Create isolated main classes for TMS and Metrics services

### PR DESCRIPTION
**Pull request by chtyim**
- Introduced cdap-master-spi for class isolation
- Created main classes for TMS and Metrics services
- Added K8s implementation of cdap-master-spi

### CODE REVIEW

File: cdap-master-spi/src/main/java/co/cask/cdap/master/spi/environment/MasterEnvironment.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+import java.util.function.Supplier;
+
+/**
+ * The interface is the integration point for CDAP master runtime provider.
+ */
+public interface MasterEnvironment {
+
+  /**
+   * This method will be invoked to initialize this {@link MasterEnvironment}.
+   * It be called before any other method is called.
```
Comment (albertshau) [inline cdap-master-spi/src/main/java/co/cask/cdap/master/spi/environment/MasterEnvironment.java]: It be -> It will be

File: cdap-master/src/main/java/co/cask/cdap/master/environment/MasterEnvironmentExtensionLoader.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+      return resources;
+    }
+
+    synchronized (this) {
+      resources = this.allowedResources;
+      if (resources != null) {
+        return resources;
+      }
+      try {
+        this.allowedResources = resources = ClassPathResources.getResourcesWithDependencies(getClass().getClassLoader(),
```
Comment (albertshau) [inline cdap-master/src/main/java/co/cask/cdap/master/environment/MasterEnvironmentExtensionLoader.java:81]: Why include all the dependencies vs just what's defined in the SPI? Probably worth a comment.

File: cdap-master/src/main/java/co/cask/cdap/master/environment/k8s/AbstractServiceMain.java [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+  /**
+   * Helper method for sub-class to call from static void main.
+   *
+   * @param mainClass the class of the master main class implementation
+   * @param args arguments to main
+   * @param <T> type of the master main class
+   * @throws Exception if execution failed
+   */
+  protected static <T extends AbstractServiceMain> void main(Class<T> mainClass, String[] args) throws Exception {
+    ClassLoader classLoader = MainClassLoader.createFromContext();
```
Comment (albertshau) [inline cdap-master/src/main/java/co/cask/cdap/master/environment/k8s/AbstractServiceMain.java:81]: why is this needed?

File: cdap-master/src/main/java/co/cask/cdap/master/environment/k8s/MetricsServiceMain.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+    MetricsContext metricsContext = injector.getInstance(MetricsCollectionService.class)
+      .getContext(Constants.Metrics.METRICS_PROCESSOR_CONTEXT);
+
+    services.add(injector.getInstance(MessagingMetricsProcessorServiceFactory.class)
+                   .create(topicNumbers, metricsContext, 0));
+    services.add(injector.getInstance(MetricsProcessorStatusService.class));
+    services.add(injector.getInstance(MetricsQueryService.class));
+  }
+  @Nullable
```
Comment (albertshau) [inline cdap-master/src/main/java/co/cask/cdap/master/environment/k8s/MetricsServiceMain.java:93]: add newline between methods

**Review by albertshau — COMMENTED** [review]
a couple questions, may just need more comments

File: cdap-master/src/main/java/co/cask/cdap/master/environment/MasterEnvironmentExtensionLoader.java (hunk 2 above)
Comment (chtyim) [reply cdap-master/src/main/java/co/cask/cdap/master/environment/MasterEnvironmentExtensionLoader.java:81]: Typically that's what it should be, as that's what classloading contract is about. The parent classloader should allow all classes loadable from the SPI module (this is dependencies of the cdap-master-spi module only, not cdap-master).

File: cdap-master/src/main/java/co/cask/cdap/master/environment/k8s/AbstractServiceMain.java (hunk 3 above)
Comment (chtyim) [reply cdap-master/src/main/java/co/cask/cdap/master/environment/k8s/AbstractServiceMain.java:81]: This is for class interception and rewriting purpose. It is used for rewriting Dataset classes as well as classes annotated with the `AuthEnforce` annotation.

All CDAP processes have to be using the MainClassLoader as the system classloader. We do the same in StandaloneMain as well as for all master TwillRunnables, although for the twill case, it is handled by twill and we only need to provide the class name of the ClassLoader to use.

File: cdap-master-spi/src/main/java/co/cask/cdap/master/spi/environment/MasterEnvironment.java (hunk 1 above)
Comment (chtyim) [reply cdap-master-spi/src/main/java/co/cask/cdap/master/spi/environment/MasterEnvironment.java]: fixed

File: cdap-master/src/main/java/co/cask/cdap/master/environment/MasterEnvironmentExtensionLoader.java (hunk 2 above)
Comment (chtyim) [reply cdap-master/src/main/java/co/cask/cdap/master/environment/MasterEnvironmentExtensionLoader.java:81]: E.g. `DiscoveryServiceClient` interface, which comes from twill-discovery-api, has to be visible to extensions, since it is the contract between the spi and extension.

File: cdap-master/src/main/java/co/cask/cdap/master/environment/k8s/MetricsServiceMain.java (hunk 4 above)
Comment (chtyim) [reply cdap-master/src/main/java/co/cask/cdap/master/environment/k8s/MetricsServiceMain.java:93]: fixed

**Review by albertshau — APPROVED** [review]
lgtm

**Comment by chtyim** [comment]
Build already passed https://builds.cask.co/browse/CDAP-DUT6688
The last commit only update comments and the RAT already passed.
Squash and merge
